### PR TITLE
feat: add message validation based on schema

### PR DESCRIPTION
**Pull request by smoya**
**Description**

This PR adds the required code to validate Kafka message against a JSON Schema document, which is grabbed from the message payload declared in the AsyncAPI document.

It introduces a new error validation notification mechanism, that allows configuring a Go channel where all validation errors will flow. Now the behavior is just logging those validation errors.

You can see a demo of this feature working [here](https://cln.sh/QZg3RT)

**Related issue(s)**
Fixes https://github.com/asyncapi/event-gateway/issues/10

### CODE REVIEW

**Review by fmvilas — APPROVED** [review]

File: asyncapi/v2/validation.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+			// Assuming there is only one message per operation as per Asyncapi 2.x.x.
+			// See https://github.com/asyncapi/event-gateway/issues/10
+			if len(o.Messages()) > 1 {
+				return nil, fmt.Errorf("can not generate message validation for operation %s. Reason: the operation has more than one message and we can't correlate which one is it", o.ID())
+			}
+
+			if len(o.Messages()) == 0 {
+				return nil, fmt.Errorf("can not generate message validation for operation %s. Reason:. Operation has no message. This is totally unexpected", o.ID())
+			}
```
Comment (magicmatatjahu) [inline asyncapi/v2/validation.go:29]: Probably I understand the problem described in the mentioned issue - this https://github.com/asyncapi/spec/issues/458 should help for you. However, can we treat the payloads for all defined messages for operation like `oneOf` for the operation? I mean, changing the shape of payload messages to:

```yaml
oneOf:
  - ... shape of payload from message1
  - ... shape of payload from message2
  - ...
```

and then you can validate all messages, but you haven't info from which message you have payload. Is it or not unacceptable in the solution?

File: proxy/validation.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	return strings.Join(errs, " | ")
+}
+
+// ValidationErrorNotifier notifies whenever a ValidationError happens.
+type ValidationErrorNotifier func(validationError *ValidationError) error
+
+// ValidationErrorToChanNotifier notifies to a given chan when a ValidationError happens.
+func ValidationErrorToChanNotifier(errChan chan *ValidationError) ValidationErrorNotifier {
+	return func(validationError *ValidationError) error {
+		// TODO Blocking or non blocking? Shall we just fire and forget via goroutine instead?
```
Comment (magicmatatjahu) [inline proxy/validation.go:51]: I don't know 😅

**Review by magicmatatjahu — COMMENTED** [review]
I have only concerns to the way of message validation. Great job! :)

File: asyncapi/v2/v2.go [hunk 3]
```diff
@@ -250,12 +257,22 @@ type SubscribeOperation struct {
 	Operation
 }
 
+// NewSubscribeOperation creates a new SubscribeOperation. Useful for testing.
+func NewSubscribeOperation(msg *Message) *SubscribeOperation {
+	return &SubscribeOperation{Operation: *NewOperation(OperationTypeSubscribe, msg)}
+}
+
```
[the remark this answers is not in the document]
Comment (smoya) [reply asyncapi/v2/v2.go:264]: Both structs are needed because of the way the decoding works. In order to set the `Type` field on the `Operation` based on the field name (Subscribe field means an Operation of type `Subscribe`, and so on) we need to set the value of that `Type` field as default. That's the purpose of the `MapStructureDefaults` method both structs implement (I created an interface called `MapStructureDefaultsProvider` for this). 

That can't be done in the `Operation` struct since once the decoder is decoding the `Operation` struct, it does not know about the field name it holds that `Operation` struct. 
In summary, whenever an `Operation` struct is being decoded, we "can't" know if it belongs to `Subscribe` or `Publish` field.

Of course we could know, but we would need to build a very customized decoder that does all the particular logic. 
My idea behind not holding the whole raw JSON document on each struct and access the values through that is for pure performance: All this unmarshaling happens on bootstrap time, however accessing the fields is in runtime, so better to have everything in place once the app runs.

Does it make sense?

File: asyncapi/v2/v2.go [hunk 4]
```diff
@@ -377,57 +407,57 @@ func (s Schemas) ToInterface(dst map[string]asyncapi.Schema) map[string]asyncapi
...
+	OneOfField                *Schema           `mapstructure:"oneOf" json:"oneOf,omitempty"`
+	PatternField              string            `mapstructure:"pattern" json:"pattern,omitempty"`
+	PatternPropertiesField    Schemas           `mapstructure:"patternProperties" json:"patternProperties,omitempty"`
+	PropertiesField           Schemas           `mapstructure:"properties" json:"properties,omitempty"`
+	PropertyNamesField        *Schema           `mapstructure:"propertyNames" json:"propertyNames,omitempty"`
+	ReadOnlyField             bool              `mapstructure:"readOnly" json:"readOnly,omitempty"`
+	RequiredField             []string          `mapstructure:"required" json:"required,omitempty"`
+	ThenField                 *Schema           `mapstructure:"then" json:"then,omitempty"`
+	TitleField                string            `mapstructure:"title" json:"title,omitempty"`
+	TypeField                 interface{}       `mapstructure:"type" json:"type"` // string | []string
```
[the remark this answers is not in the document]
Comment (smoya) [reply asyncapi/v2/v2.go:452]: TIL :) Changing it right now. Thanks! 🕺

File: proxy/validation.go (hunk 2 above)
Comment (smoya) [reply proxy/validation.go:51]: TODOS in code are not always a bad thing :)

File: asyncapi/v2/validation.go (hunk 1 above)
Comment (smoya) [reply asyncapi/v2/validation.go:29]: Good point. it also makes me realize `operation.Messages()` should check for the `oneOf` as well right?

Just to be aligned: you are suggesting that whenever `operation.Messages()` returns more than 1 message, we create a Schema with `oneOf` containing all of those messages and use it for validating. Am I right?

So `operation.Messages()` will return each message individually, but then for making the JSON Schema validation, all of those will be added into just one Schema with a `oneOf`.

File: asyncapi/v2/v2.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply asyncapi/v2/v2.go:264]: It make sense, so TLD;DR; In Golang simpler is decoding object as:

```bash
`subscribe` -> SubscribeOperation
`publish` -> PublishOperation
```

than

```
`subscribe` -> Operation with `type` field set to `subscribe`
`publish` -> Operation with `type` field set to `publish`
```

Am I right?

[the remark this answers is not in the document]
Comment (smoya) [reply asyncapi/v2/v2.go:264]: Yeah, you are right. The point is that I can't tell the decoder to use a default value for the `type` field at this moment. At least the decoders (both json and mapstructure) don't support default values or setting constructors as a way of generating the object.

File: asyncapi/v2/validation.go (hunk 1 above)
Comment (smoya) [reply asyncapi/v2/validation.go:29]: Nah, nvm. I'm wrong with my assumption. There is only one message but several possible payloads. Is not the same.

Comment (magicmatatjahu) [reply asyncapi/v2/validation.go:29]: > Good point. it also makes me realize operation.Messages() should check for the oneOf as well right?

In AsyncAPI we can have construct like:

```yaml
message: 
  payload: ...
  
# or multiple messages

message:
  oneOf:
     - ...
     - ...
```

so `operation.Messages` should read the `oneOf` keyword.

> Nah, nvm. I'm wrong with my assumption. There is only one message but several possible payloads. Is not the same. I think the way it's done right now it's right, isn't?

Yes, for multiple payloads now is ok, but I thought about merging the payloads from messages by `oneOf` keywords, so if you have this:

```yaml
message:
  oneOf:
     - payload: ... # payload1
     - payload: ... # payload2
```

we can merge to:

```yaml
oneOf:
  - ...payload1 
  - ...payload2
```

and then validate all payload from all messages :)

Comment (smoya) [reply asyncapi/v2/validation.go:29]: But there is only one message per operation. I'm not sure I'm following you. 
If I'm not wrong, `operation.Messages()` should always return one message (technically an array but with only one message) in `v2.x.x`, right? Meaning, the payload of such message will be a Schema that will contain that `oneOf` already and will always validate accordingly.

File: asyncapi/v2/v2.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (smoya) [reply asyncapi/v2/v2.go]: done

File: asyncapi/v2/validation.go (hunk 1 above)
Comment (magicmatatjahu) [reply asyncapi/v2/validation.go:29]: > If I'm not wrong, operation.Messages() should always return one message (technically an array but with only one message) in v2.x.x, right?

Nope :) In `v2.X.X` you can define multiple messages for given operation by construct:

```yaml
message:
  oneOf:
     - payload: ... # payload1
     - payload: ... # payload2
```

So `operation.Messages()` functions should return the array of messages if there is an `oneOf` keyword, from the [spec](https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md#fixed-fields-7):

> A definition of the message that will be published or received on this channel. oneOf is allowed here to specify multiple messages, however, a message MUST be valid only against one of the referenced message objects.

Check `oneOf` mention :)

> Maybe I understand you wrong.

My idea is based on merging only the payloads, when operation has defined multiple messages (and each message has payload field), so validation would be performed on all payloads, but concatenated with `oneOf` keyword. This concatenation should be done in the code, something like:

```go
if len(o.Messages()) > 1 {
  // create new object with concatenating payloads from messages by `oneOf` keyword
}
```

Comment (magicmatatjahu) [reply asyncapi/v2/validation.go:29]: However, we cannot infer which message arrives to the gateway (we don't have such a `messageId` field) so your current solution where you are handling only one message is enough.

File: asyncapi/v2/v2.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply asyncapi/v2/v2.go:264]: Supporting `discriminators` in golang will be pain 😄  But yeah, I understand now, it can stay.

File: asyncapi/v2/validation.go (hunk 1 above)
Comment (smoya) [reply asyncapi/v2/validation.go]: The current solution works. However is not complete since `oneOf` should take into account (even though not extrictly necessary for the event gateway right now).
As the idea is to eventually move all the `asyncapi` package to the `parser-go` repository, I did a small iteration on this code fixing the `oneOf` stuff.

I also fixed the `additionalProperties` and `additionalItems` fields so they are now variadic (either `false` or `*Schema`) so I needed the first for testing.

Comment (smoya) [reply asyncapi/v2/validation.go]: Btw, all fixes on the document interface are being fixed in the `parser-api` repository here: https://github.com/asyncapi/parser-api/pull/10

File: asyncapi/document.go [hunk 5]
```diff
@@ -80,14 +80,22 @@ type Message interface {
 	Payload() Schema
 }
 
+// FalsifiableSchema is a variadic type used for some Schema fields.
+// For example, additionalProperties value can be either `false` or a Schema.
+type FalsifiableSchema interface {
+	IsFalse() bool
+	IsSchema() bool
+	Schema() Schema
+}
```
Comment (magicmatatjahu) [inline asyncapi/document.go:89]: Sorry if I'm being disruptive with this review 😅   I think that `FalsifiableSchema` isn't correct solution. Of course, `additionalProperties` and `additionalItems` have option to be `false`, but also have option to be `true` :) Also, the JSON Schema can be either a boolean or Schema object (in places where you can use the Schema, like in `items` or `properties`), so I think that:

```go
type BooleanbleSchema interface {
        isBoolean() bool
	Boolean() bool
        IsSchema() bool
	Schema() Schema
}
```

should be better to handle Schema, true and false :) - like we have in our ParserJS. https://github.com/asyncapi/parser-js/blob/master/lib/models/schema.js#L368

File: asyncapi/v2/validation.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+			var payload asyncapi.Schema
+			var messageNames string
+			if len(o.Messages()) > 1 {
+				// Meaning message payload is a Schema containing several payloads as `oneOf`.
+				// Generating back just one Schema adding all payloads to oneOf field.
```
Comment (magicmatatjahu) [inline asyncapi/v2/validation.go:30]: But be aware that if we will have support for `messageID` then it should be changed to single payload :) Maybe add this info in the comment?

**Review by magicmatatjahu — COMMENTED** [review]
Two more comments 😅  But overall good!

File: asyncapi/v2/validation.go (hunk 6 above)
Comment (smoya) [reply asyncapi/v2/validation.go:30]: won't `messageID` be, if so, in version 3.x.x of the spec? I know it does not break anything anyway.

Comment (magicmatatjahu) [reply asyncapi/v2/validation.go:30]: Probably in the `3.X.X` or maybe in the `2.2.X`. It doesn't break, but as we had discussion, you merge here the payloads from all messages (from `oneOf`), so at the moment you haven't information which message do you handle - https://github.com/asyncapi/event-gateway/issues/10 :)

Comment (smoya) [reply asyncapi/v2/validation.go:30]: Once we add messageID, we will need to add support obviously. So I think it's a thing of the future, a future that it is not so close as https://github.com/asyncapi/spec/issues/458 is in RFC1 status. 
https://github.com/asyncapi/parser-go is not even supporting `2.1.0` yet.

File: asyncapi/document.go (hunk 5 above)
Comment (smoya) [reply asyncapi/document.go:89]: No you are not! This is the kind of review I'm looking for. You have a lot of experience with JSON Schema, a thing I have not :) 
I think we can move forward with this solution on this PR. I opened an issue here https://github.com/asyncapi/event-gateway/issues/22 which I'll work next on it for improving the whole Schema thing, including this boolean behavior you mentioned here.

File: asyncapi/v2/v2.go [hunk 7]
```diff
@@ -297,8 +327,21 @@ func (o Operation) IsClientSubscribing() bool {
...
+
+		var msgs []asyncapi.Message
+		for _, payload := range o.MessageField.Payload().OneOf() {
+			p := payload.(*Schema)
+			msgs = append(msgs, Message{
+				PayloadField: p,
+			})
+		}
+
+		return msgs
```
Comment (magicmatatjahu) [inline asyncapi/v2/v2.go:342]: It's not the true. `MessageField` should have `OneOf` property and you should use it, not `Payload.OneOf()`. `Payload.OneOF` is related to single message, not to array of messages :)

File: asyncapi/document.go (hunk 5 above)
Comment (magicmatatjahu) [reply asyncapi/document.go:89]: Ok :)

File: asyncapi/v2/v2.go (hunk 7 above)
Comment (smoya) [reply asyncapi/v2/v2.go:342]: Wow, this is not explained at all in the documentation :O Well, only through an example but not clear at all: https://www.asyncapi.com/docs/specifications/v2.0.0#messageObject

Is there any other exception or hidden gem that you know about that I should add here?

Comment (magicmatatjahu) [reply asyncapi/v2/v2.go:342]: You have information about `oneOf` in the https://www.asyncapi.com/docs/specifications/v2.0.0#operationObject

for the message field, you can read:

> A definition of the message that will be published or received on this channel. oneOf is allowed here to specify multiple messages, however, a message MUST be valid only against one of the referenced message objects.

so you can have shape:

```yaml
publish:
  message
    payload: ...
```

or

```yaml
publish:
  message:
    oneOf:
      - payload: 
      - payload: 
```

Comment (smoya) [reply asyncapi/v2/v2.go:342]: Yeah, This `oneOf` field should be either part of the message object or a new type that holds that field.
Ok, gotta work on it! thks!

Comment (smoya) [reply asyncapi/v2/v2.go]: This is done

**Review by magicmatatjahu — APPROVED** [review]
LGTM! 🚀

File: asyncapi/v2/v2.go (hunk 7 above)
Comment (smoya) [reply asyncapi/v2/v2.go]: For reference https://github.com/asyncapi/spec/issues/588
